hierarqcal/core.py

```diff
@@ -636,23 +650,241 @@ def __eq__(self, other):
+        Each time a motif in the stack changes, a loop runs through the stack from the beginning and calls each motif to update the graph (the available qubits, the edges etc).
+
+        Args:
+            Qc_l (list): List of available qubits.
+            *args: Variable length argument list.
+            **kwargs: Arbitrary keyword arguments, such as:
+                * mapping (tuple(function, int)): TODO update this docstring, it's not a tuple anymore.
+                    Function mapping is specified as a tuple, where the first argument is a function and the second is the number of symbols it uses. A symbol here refers to an variational paramater for a quantum circuit, i.e. crz(theta, q0, q1) <- theta is a symbol for the gate.
+
+        Returns:
+            Qconv: Returns the updated version of itself, with correct nodes and edges.
+        """
+        n_pivot = len([x for x in self.local_pattern if x == "1"])
+
+        if len(self.local_pattern.replace("*","")) > self.arity:
+                    # TODO explain this...
+                    self.local_pattern == "*1"
+
+        l_pattern = self.wildcard_populate(self.local_pattern, self.arity) 
+
+        
+
+        if len(self.pattern.replace("1", "1" * n_pivot).replace("*","")) > len(Qc_l):
+            # TODO explain this...
+            print(self.pattern.replace("1", "1" * n_pivot), len(self.pattern.replace("1", "1" * n_pivot)), len(Qc_l))
+            raise ValueError(
+                "Pattern, taking into account the local pattern, is longer than available qubits, please use a shorter pattern or more qubits"
+            )
+
+        self.pivot_pattern_fn = self.get_pivot_pattern_fn(
+            self.pattern.replace("1", "1" * n_pivot), Qc_l
+        )
+
+        # group every n_pivot elements of self.pivot_pattern_fn(Qc_l) into a tuple
+        pivot_q = [
+            tuple(self.pivot_pattern_fn(Qc_l)[i * n_pivot : (i + 1) * n_pivot])
+            for i in range((len(self.pivot_pattern_fn(Qc_l)) + n_pivot - 1) // n_pivot)
+        ]
+
+        remaining_q = [q for q in Qc_l if q not in [p for P in pivot_q for p in P]]
+
+        if self.arity > 1:
+            nq_available = len(remaining_q)
+
+            if nq_available > 0:
+
+                if self.stride % nq_available == 0:
+                    # TODO make this clear in documentation
+                    # warnings.warn(
+                    #     f"Stride and number of available qubits can't be the same, received:\nstride: {self.stride}\n available qubits:{nq_available}. Defaulting to stride of 1"
+                    # )
+                    self.stride = 1
+               
+                r_q = remaining_q[self.offset :: self.step]
+                if self.boundary == "periodic":
+                    while len(r_q) < len(remaining_q):
+                        r_q += [q for q in remaining_q if q not in r_q][:: self.step]
+                Ec_l = []
+                while len(r_q) > 0:
```

Comment:
As a safety measure, can we add a max_it type check to ensure both these while loops terminate.

Here's some (not sanitary) inputs that runs forever:
```python
N = 10
u = Qunitary("cnot()^10;cnot()^01")
boundary = "open"
qft = Qinit(N) + Qpivot("1*", local_pattern="11*", offset=0, mapping=u, boundary=boundary)
circuit = qft(backend="qiskit")
circuit.draw("mpl")

N = 10
u = Qunitary("cnot()^10;")
boundary = "open"
qft = Qinit(N) + Qpivot("1*", local_pattern="11*", offset=0, mapping=u, boundary=boundary)
circuit = qft(backend="qiskit")
circuit.draw("mpl")
```
Btw this is really awesome, nice!:
```python
N = 10
u = Qunitary("cnot()^20;cnot()^21")
boundary = "open"
qft = Qinit(N) + Qpivot("1*", local_pattern="11*", offset=0, mapping=u, boundary=boundary)
circuit = qft(backend="qiskit")
circuit.draw("mpl")
```